cmake/modules/extensions.cmake

```diff
@@ -117,6 +117,10 @@ function(zephyr_libc_link_libraries)
   set_property(TARGET zephyr_interface APPEND PROPERTY LIBC_LINK_LIBRARIES ${ARGV})
 endfunction()
 
+function(zephyr_kernel_link_libraries)
+  set_property(TARGET zephyr_interface APPEND PROPERTY KERNEL_LINK_LIBRARIES ${ARGV})
+endfunction()
```

Comment:
The more I think about this, the more I have second thoughts.
Here are some of my concerns:
- Using the zephyr_interface to pass information to the kernel lib seems a bit wrong, and only reason I can see for the is because `zephyr_interface` happens to be defined early, whereas `kernel` target is not.
- Using a custom property prevents the use of generator expressions, which could be useful in some cases
- Exposing `zephyr_kernel_link_libraries` to the public bears an additional risk of misuse. We both know the distinction between zephyr and kernel, but a lot of less competent users may not, and consider that Zephyr == kernel, and therefore they might wrongly start to use `zephyr_kernel_link_libraries` when they should use `zephyr_link_libraries`.

As the function is only used internally in the `lib/libc` and for C library linking, then I would prefer a `target_link_library(kernel <c-lib>)`.
Something which would also mean correct transfer of includes, compile options, etc. based on public / interface settings of the lib being linked against. In short, regular CMake behavior.

I acknowledge such a request requires that kernel lib is defined earlier in the CMake process, but in fact we need to do something related to kernel lib anyway, which will be described in other comment.

Reply:
That sounds pretty simple then ; probably just have the top-level CMakeLists.txt file define `kernel` early in the process. It's not like we won't always need that one :-)

Reply:
That took a bit of a kludge -- we need to know at the time of `add_library` whether we're adding a real library or just an INTERFACE. I moved the logic for figuring that out into the top-level CMakeLists.txt and carried that data along in a property on the kernel target for use in kernel/CMakeLists.txt. Otherwise, this seems like a reasonable cleanup.